### PR TITLE
feat: Misc updates to zone timer entities

### PR DESCRIPTION
**Pull request by Panzerhandschuh**
- Added checkpoints ordered field to zone_timer_start/stage
- Limit ground speed enabled by default
- Filter names now use filterclass dropdown
- Remove "Stages Ordered" property from zone_timer_start
- Remove zone_timer_stage_end
- Add "Stage End Zones" property with enum values: "Next Stage", "Last Checkpoint"

### CODE REVIEW

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (Panzerhandschuh) [inline fgd/brush/trigger/zone_timer_stage.fgd:8]: @jason-e Let me know if the wording here makes sense. If the checkpoints are optional, then the player shouldn't have to activate the checkpoints in order to complete the stage.

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (Panzerhandschuh) [inline fgd/brush/trigger/zone_timer_stage.fgd:9]: I defaulted this to true because I think it is more common to have stages that prevent bhopping.

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (Panzerhandschuh) [inline fgd/brush/trigger/zone_timer_stage.fgd:7]: I think checkpoints should only be required if they are a critical objective in the map (ex: bhop_floodwaters). Keeping them optional gives the player more freedom to find interesting sequence breaks (ex: https://www.youtube.com/watch?v=gJTNid3Hzgg&t=43s).

**Comment by Panzerhandschuh** [comment]
I'll be making some changes based on the stage zone discussion from last night:
- Remove "Stages Ordered" property from zone_timer_start
- Remove zone_timer_stage_end
- Add "Stages End at Stage Starts"? property to zone_timer_start

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (jason-e) [reply fgd/brush/trigger/zone_timer_stage.fgd:7]: Elaborating on my thoughts from earlier this week:

1. I think we are working towards timing/split consistency in general, and so we should be specifically promoting robust checkpoint placement that isn't dependent on routing. We would rather a mapper put in the effort to make sure checkpoints are unskippable by placing them well or altering the map (for new maps). Or to look at it another way, it seems weird to explicitly codify an opening for user error.
2. The default we choose here is not necessarily the same as enforcing whatever philosophy we do go with. Independent users are going to set this based on their own interpretation, and it's hard to say for sure but I would think they are going to tend to submit maps with this turned on unless they explicitly want it off. So, if we do want to promote a certain mentality, that's really something we will do when reviewing submitted zones.

The surf_treespam example is not a great case for this idea because that route skips one ramp in a spot where one would never decide to put a checkpoint. Skips aside, checkpoints should be placed at clear chokepoints when at all possible to minimize timing differences due to insignificant route differences.

Another angle on this is that unintended skips are not always desirable, and this is especially true for ones that skip a checkpoint zone, which could involve skipping a large section of a map. If someone finds an unintended skip that's significant enough to skip a checkpoint, I would rather they have to bring it up with us so we can decide to enable or not, instead of it just working right way leaving us to clean up the mess if it's undesirable (i.e. someone finds it and uses it but doesn't tell us because they know it might be excessive, then other people copy them, then when we finally get wind of it we have to remove a bunch of runs from people who assumed it was condoned -- I've seen this happen too many times).

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (jason-e) [reply fgd/brush/trigger/zone_timer_stage.fgd:8]: Maybe:
"Do the checkpoints in this segment have a logical order, even if the player is not required to activate them all?

Could also add to elaborate even further to drive it home (your call):
"If an ordered checkpoint is skipped by activating a higher numbered checkpoint first, the lower-numbered checkpoint cannot be activated later in the run."

File: fgd/brush/trigger/zone_timer_stage.fgd
Comment (Panzerhandschuh) [reply fgd/brush/trigger/zone_timer_stage.fgd:7]: Both surf_fast and surf_classics2 have skips that I'm not too fond of, but checkpoints probably wouldn't have prevented them since they're staged maps:
https://youtu.be/8Md12-c0EUk?si=O_CzFoC6VR3KLf5_&t=6
https://youtu.be/4yx0QQLh4S0?si=HDplcWvEgmQUJs3R&t=140

Even if these skips were on linear maps, the mapper might happen to place their checkpoints before and after the section that can be skipped. Checkpoints can't be fully relied on to make the player follow an intended route. I think progression logic should be explicitly enforced by progress teleports, filters, and collectibles. What happens when a player skips a collectible but hits all required checkpoints? It's an unlikely scenario but it's why I think progression should be coupled with the entities that grant progress, and not the checkpoint system. Checkpoint skipping will probably come up as an issue on bhop maps more so than surf and I still don't know what the UX will be for when the player misses a checkpoint.

I think I'm in a minority position on this. I'll discuss it with the team on Saturday and go with whatever the majority of the team wants.

Comment (jason-e) [reply fgd/brush/trigger/zone_timer_stage.fgd:7]: > Checkpoints can't be fully relied on to make the player follow an intended route

True, but it can still *help* enforce how a map is played, even if it isn't bulletproof on its own

> Progression logic should be explicitly enforced by progress teleports, filters, and collectibles.

It seems like this issue might be more about the role of checkpoints than skip philosophy. If we *were* to make checkpoints themselves always optional, there's nothing stopping a mapper from making them required anyway with a complicated sequence of filters, in which case all we will have done is make it more difficult for the mapper to make the map do what they want. If they think the map seems "unskippable" so they doesn't bother doing this (because it's annoying to set up) but then a skip is found anyway, we've really just tricked them into accidentally leaking a skip into the wild.

We could theoretically make gating progress even easier with new entities, but at that point we would just be splitting up required checkpoint zones into two components that must be placed separately when we really want them to be linked (making explicit progress in a map is a good time to split the timer).

> What happens when a player skips a collectible but hits all required checkpoints?

Not sure I'm following here -- if you're talking about a map with required+unordered checkpoints at collectible locations, they should either have a checkpoint corresponding to every collectible (in which case they don't need to use the collectible system at all), and/or should physically gate the progress of the player with a filter_momentum_collectibles that requires all of them (or even better, in Momentum mappers could open a solid door when all collectibles are gathered)

> Checkpoint skipping will probably come up as an issue on bhop maps

Bhop definitely has more of a culture of wacky routes and unexpected skips, so that does raise the likelihood of non-required checkpoints making sense for any given map (because it's still a valid *option*, don't get me wrong), but we should still promote good checkpoint placement that avoids the problem entirely because skippable splits are less useful than ones everyone hits.

Comment (Panzerhandschuh) [reply fgd/brush/trigger/zone_timer_stage.fgd:7]: You should probably listen to the meeting recording regarding this change. The team leaned towards keeping the checkpoints optional by default. Apparently this is expected in RJ and SJ and the modes I've mentioned before (bhop, defrag, and ahop). Surf might be the only mode where required checkpoints would be preferred by default. Feel free to debate the team further if you'd like. I don't care too much either way since there are good arguments for either side.

**Comment by Panzerhandschuh** [comment]
Don't merge this until rio's zone def changes are ready.

**Review by Gocnak — APPROVED** [review]

**Review by jason-e — APPROVED** [review]

**Pull request merged**
